File: 0787. Cheapest Flights Within K Stops - Medium/solution.cpp
static auto _ = [] () { ios_base::sync_with_stdio(false); cin.tie(nullptr); return 0; } ();
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int K) {
        vector < int > ford_bellman(n, 1e8);
        ford_bellman[src] = 0;
        
        for(int i = 0; i <= K; i++){
            vector < int > updated(ford_bellman);
            for(const auto& edje: flights) {
                updated[edje[1]] = min(updated[edje[1]], ford_bellman[edje[0]] + edje[2]);
            }
            ford_bellman = move(updated);
        }
        return ford_bellman[dst] == 1e8 ? -1 : ford_bellman[dst];
    }
};

File: 0787. Cheapest Flights Within K Stops - Medium/solution.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, K: int) -> int:
        graph = collections.defaultdict(list)
        q = collections.deque()
        min_price = float('inf')

        for u, v, w in flights:
            graph[u].append((w, v))

        q.append((src, 0, 0))
        while q:
            city, stops, price = q.popleft()
            if city == dst:
                min_price = min(min_price, price)
                continue

            if stops <= K and price <= min_price:
                for price_to_neighbour, neighbour in graph[city]:
                    q.append((neighbour, stops+1, price+price_to_neighbour))

        return min_price if min_price!=float('inf') else -1